### PR TITLE
[Frontend] Add support for while loops in autograph

### PR DESCRIPTION
**Pull request by grwlf**
Add support for while-loops in the `@qjit(autograph=True)` mode

We replace the while-loops with `catalyst.while_loop` calls on a best-efforts basis. In case of errors we fallback to the unrolled loop path. A decision was made to not warn by default that the fallback has been used. 

[sc-41300]

### CODE REVIEW

File: frontend/catalyst/ag_primitives.py [hunk 1]
```diff
@@ -106,7 +107,9 @@ def functional_cond():
     # manipulate tuples when they don't expect it. Ensure set_state receives a tuple regardless.
     results = functional_cond()
     if not isinstance(results, tuple):
-        results = (results,)
+        # TODO: Do we have a test-case for this?
+        results = (results,)  # pragma: nocover
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:111]: Hmm, I definitely encountered that issue during development, but maybe there is no test case for it 🤔

**Review by dime10 — COMMENTED** [review]
Thanks @grwlf, looking good so far 🚀
Some preliminary comments:

**Comment by dime10** [comment]
Oh, this PR will also need to update the newly added autograph docs :)

**Comment by josh146** [comment]
> Oh, this PR will also need to update the newly added autograph docs :)

I agree, but in this case might argue for doing it as a separate PR - just because the docs still pin to `latest`, so readers will be seeing support for something that they can't pip install.

**Comment by dime10** [comment]
> I agree, but in this case might argue for doing it as a separate PR - just because the docs still pin to latest, so readers will be seeing support for something that they can't pip install.

This is tricky, because we can't prevent new functions from not showing up in the docs either 🤔 (well we probably could but ...)
I think that's why I originally wanted the docs to show the latest release.
@josh146 Do you think it would make sense to have a branch that's identical to the last release, but can contain corrections (or additions like the autograph docs), and have the docs show that branch by default?

File: frontend/catalyst/__init__.py [hunk 2]
```diff
@@ -76,6 +76,7 @@
 )
 from catalyst.utils.exceptions import CompileError
 
+autograph_force_fallbacks = False
```
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/__init__.py:79]: I want it exclusively for testing. The idea is to set this variable and test that the fallback parts of autograph work.

File: mlir/lib/Driver/CompilerDriver.cpp
[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:1]: Oh, there is another [PR](https://github.com/PennyLaneAI/catalyst/pull/317) for that, I should have mentioned this..

File: frontend/catalyst/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/__init__.py:79]: Interesting. These global flags are going to be removed in the near future to be replaced with a qjit parameter for the autograph mode, with 3 options:
- strict: no fallbacks allowed, if there is an array during the conversion the error is immediately raised to the user
- permissive (default): autograph will always fall back to Python if it can't convert the user code, but in some instances it will raise a warning to allow the user to adjust their code to be more optimal
- silent: same as permissive, but no warnings are raised

I'm not sure how your new flag would fit into this 🤔 @josh146 any ideas?

File: mlir/lib/Driver/CompilerDriver.cpp
[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Driver/CompilerDriver.cpp:1]: So the changes can be removed from this PR?

File: frontend/catalyst/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/__init__.py:79]: We could have `autograph_mode="force-fallback"`? I'm agnostic here -- as long as there are places where this would be useful, and there are no other ways to solve @grwlf's testing problem, it could be worth supporting.

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/__init__.py:79]: @grwlf Could your testing be solved differently?

File: mlir/lib/Driver/CompilerDriver.cpp
[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:1]: I have tried to change the base of this PR but it only got worse so I changed it back. I think we can merge the other PR first, and then solve this one. What do you think?

[the remark this answers is not in the document]
Comment (dime10) [reply mlir/lib/Driver/CompilerDriver.cpp:1]: Sounds good 👍

File: frontend/catalyst/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/__init__.py:79]: Hmm, I think the current version is kind of less invasive than others I see (like, triggering an artificial error forcing autograph to fallback), I will think.

[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/__init__.py]: So I moved the test flag out from the `__init__` into the `ag_primitives` and changed its semantics. @dime10 could you please check?

File: frontend/catalyst/ag_primitives.py (hunk 1 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:114]: I think the only otrside entity that affects `result` is catalyst.for_loop. We ask it to always return `get_state` which is always a tuple, so I think it should be ok to remove the precaution. Did you do the development before the pytree support was added into catalyst ?

File: frontend/catalyst/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (josh146) [reply frontend/catalyst/__init__.py]: @grwlf thinking about this overnight -- if this is something that is primarily for development testing, and doesn't have as much value to be user-facing, then it implies it might be best as a custom [test fixture](https://docs.pytest.org/en/6.2.x/fixture.html).

That is, a test fixture that sets up a patch to force this behaviour, and then has a tear down. You could include this fixture on the TestClass that requires this behaviour.

File: mlir/lib/Driver/CompilerDriver.cpp
[the remark this answers is not in the document]
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp:1]: Other PR is merged, no more CompileDriver changes in this one.

**Comment by dime10** [comment]
@grwlf Can you open a PR that updates the autograph docs (to be merged later)?

File: frontend/catalyst/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grwlf) [reply frontend/catalyst/__init__.py]: @josh146 this `monkeypatch` thing that we are using [here](https://github.com/PennyLaneAI/catalyst/blob/6544da88f9d28be42a94a079f8969c204ee9f9fa/frontend/test/pytest/test_autograph.py#L1103) is a fixture, the built-in one. The problem is how to link it to our main code. As I see, Autograph requires us to use functions and we want to trigger the fallback path in them, so we must either use effects (here - the global variable I added) or pass arguments triggering errors on the default path. I think that it would be easier to maintain the former approach, but I am open to suggestions.

**Comment by grwlf** [comment]
> @grwlf Can you open a PR that updates the autograph docs (to be merged later)?

@dime10 here is the doc PR https://github.com/PennyLaneAI/catalyst/pull/330

File: frontend/catalyst/ag_primitives.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:114]: If the current tests pass without it, I think we can remove them

File: frontend/catalyst/ag_primitives.py [hunk 3]
```diff
@@ -318,7 +327,87 @@ def for_stmt(
...
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
+
+    if fallback:
+        results = _call_python_while(
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:407]: I think it's safer to reset the state after a failed tracing attempt before falling back to Python.

File: frontend/catalyst/ag_primitives.py [hunk 4]
```diff
@@ -318,7 +327,87 @@ def for_stmt(
...
+                f"\n"
+                f"The error ocurred within the body of the following for loop statement:\n"
+                f"{while_loop_info}"
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:404]: Do you think the warning is actually to users, at least in some instances?

File: frontend/catalyst/ag_primitives.py [hunk 5]
```diff
@@ -166,6 +172,9 @@ def _call_catalyst_for(
 ):
     """Dispatch to a Catalyst implementation of for loops."""
 
+    if _emulate_fallback_errors:
+        raise AutoGraphError("Emulated autograph fallback errror")
```
Comment (dime10) [inline frontend/catalyst/ag_primitives.py:176]: I'd like us to find a way to remove the emulation of errors from production code. If we followed this strategy the code base would be littered with such flags.

File: doc/changelog.md [hunk 6]
```diff
@@ -8,6 +8,9 @@
   messages and includes a verbose trace if verbose mode is enabled.
   [(#303)](https://github.com/PennyLaneAI/catalyst/pull/303)
 
+* The AutoGraph feature, still experimental, now supports native Python `while` loops as well.
+  [(#318)](https://github.com/PennyLaneAI/catalyst/pull/318)
```
Comment (dime10) [inline doc/changelog.md:12]: Can you add an example? :)

File: frontend/test/pytest/test_autograph.py [hunk 7]
```diff
@@ -1077,10 +1079,205 @@ def f():
...
+
+        def f1():
+            acc = 0
+            while acc < 5:
+                acc = acc + 1
+            return acc
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error")
+            qjit(autograph=True)(f1)()
```
Comment (dime10) [inline frontend/test/pytest/test_autograph.py:1241]: What does this do?

File: frontend/catalyst/ag_primitives.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:404]: Basically, the warning exists in the for loop so users can fix their code to make the conversion succeed (see for loop test cases). If we can't think of any such case for the while loop, we shouldn't raise a warning because it is not actionable for the used, and they are more likely to just silence all warnings.

**Review by dime10 — CHANGES_REQUESTED** [review]
Thanks @grwlf, just a few points to resolve and we can merge this 👍

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:176]: I see your point, but do you think we need fallback here at all?

File: frontend/catalyst/ag_primitives.py (hunk 4 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:404]: I think we need to keep the behavior observable anyways. Inability to insert `@cond_while` would almost certainly lead to code expansion which is something we should want to communicate. It might be useful for us as well, e.g. in issue reports. The contents of the warning might be different though. Let me know what do you think.

File: doc/changelog.md (hunk 6 above)
Comment (grwlf) [reply doc/changelog.md:12]: Done

File: frontend/catalyst/ag_primitives.py (hunk 3 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:409]: Indeed, done

File: frontend/catalyst/ag_primitives.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:394]: Removing the fallback does not seem sensible if we're trying to build a robust system.

> Inability to insert @while_loop would almost certainly lead to code expansion which is something we should want to communicate

Not necessarily, there is going to be loops that are not convertible (and maybe weren't intended to be). The default approach is to perform conversion on a best effort basis, allowing unconvertible constructs to execute in Python instead. This leads to an overall good experience that "just works" (as best as we can achieve this). Always raising warnings whenever this happens would be a pain point for users.

I agree that logging this type of information could be important however, for this purpose autograph has a global debug flag with different levels. We should think about this, but it would be a separate task I think.

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:170]: I tried to run your example but it fails for a different reason:
```py
/home/dave/catalyst/frontend/catalyst/ag_primitives.py:382: UserWarning: Tracing of an AutoGraph converted while loop failed with an exception:
  TypeError:    'Undefined' object is not callable
```

Comment (dime10) [reply frontend/catalyst/ag_primitives.py:170]: If this works I'm ok with it, although a realistic test case would be better (unless we can can't come up with one) 👍

File: frontend/catalyst/ag_primitives.py (hunk 4 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:394]: > Always raising warnings whenever this happens would be a pain point for users.

Not sure about the amount of pain (You might be right, I just don't know. Isn't it a painful to understand the cause of source expansion after hours of debugging?). I still think that removing this warning would make important information invisible or less visible, depending on the autograph logging capabilities you mentioned below. 

> I agree that logging this type of information could be important however, for this purpose autograph has a global debug flag with different levels. We should think about this, but it would be a separate task I think.

I also relied on this warning in testing. Without it I think I need to replace some tests with lit-tests on JAX sources. It might be useful by itself.

I still don't see why the absence of this warning is consistent with what we have for for-loops (and our logical-and-or approach looks to me also different from both these cases as well). I think it is a management/UI decision in the end, so please let me know what implementation do you want to see here.

File: frontend/catalyst/ag_primitives.py (hunk 5 above)
Comment (grwlf) [reply frontend/catalyst/ag_primitives.py:170]: Yes, I just wrote it here as a draft, I will check it more carefully.

Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: Should be implemented by now, @dime10 please check

File: doc/changelog.md [hunk 8]
```diff
@@ -8,6 +8,31 @@
...
+  Third party `QuantumDevice` implementations can now be loaded into the runtime.
+  [(#327)](https://github.com/PennyLaneAI/catalyst/pull/327)
+
+<h3>Breaking changes</h3>
+
+* The axis ordering for `catalyst.jacobian` is updated to match `jax.jacobian`. Assume we have parameters of shape 
+  `[a,b]` and results of shape `[c,d]`. The jacobian would get the shape `[c,d,a,b]` instead of `[a,b,c,d]`.
+  [(#283)](https://github.com/PennyLaneAI/catalyst/pull/283)
+
+<h3>Bug fixes</h3>
```
Comment (dime10) [inline doc/changelog.md:35]: Probably a merge mixup :)
```suggestion
```

File: frontend/catalyst/ag_primitives.py (hunk 4 above)
Comment (dime10) [reply frontend/catalyst/ag_primitives.py:385]: > Not sure about the amount of pain (You might be right, but there are sociological questions for which I don't know answers: Isn't it a painful to understand the cause of source expansion after hours of debugging? What scenario would lead to a more negative user reaction?) I only think that removing this warning might make important information invisible.

Well, I can only tell you that this is Josh's policy regarding warnings, and I'm inclined to agree with it. Warnings that can't be acted upon aren't all that useful.

> I still don't see why the absence of this warning is consistent with what we have for for-loops (and our logical-and-or approach looks to me also different from both these cases as well). I think it is a management/UI decision in the end, so if you are not convinced please let me know and I will remove it.

There is a difference. In the for loop case there are *concrete instances* that we know will fail conversion but are fixable by the user. Raising a warning then allows them to fix it.
There are are also instances where the user can't do anything to fix it and we raise the warning. This is bad, but we accept it because we can't differentiate between the fixable and unfixable case. To mitigate this, we added the configuration to allow users to silence the warnings.
Also note that there are [fallbacks in the for loop case](https://github.com/PennyLaneAI/catalyst/blob/445b15881d71fb454b05782b5d55d8526272a82d/frontend/catalyst/ag_primitives.py#L247-L262) that we don't raise a warning for, precisely because these aren't instances the user could fix, and were probably meant as Python loops in the first place.
In the case of while loop, I don't know any instance that fails conversion that the user could correct.

In short, we use warnings only sparingly and when there is immediate action the user can take to resolve the warning.

What you propose has value in a debug or verbose mode, and we should look into that, but using warnings for it doesn't match the experience we want users to have.

Comment (grwlf) [reply frontend/catalyst/ag_primitives.py]: As discussed, I have removed the warning.

**Review by dime10 — APPROVED** [review]
Great work 💯
